layout/layout.c

```diff
@@ -810,44 +810,38 @@ M0_INTERNAL uint64_t m0_layout_find_by_buffsize(struct m0_layout_domain *dom,
 						struct m0_fid *pver,
 						size_t buffsize)
 {
-	uint64_t lid = M0_DEFAULT_LAYOUT_ID;
-	struct m0_pdclust_attr *pa;
-	struct m0_layout *l;
-	uint64_t hash;
+	struct m0_pdclust_attr *pa = NULL;
+	struct m0_layout *l = NULL;
+	uint64_t hash = 0;
 	int i;
 
 	m0_mutex_lock(&dom->ld_lock);
 	for (i = M0_DEFAULT_LAYOUT_ID; i < m0_lid_to_unit_map_nr; ++i) {
-		/* Current BE max tx size constraints. */
-		if (m0_lid_to_unit_map[i] > 4 * LNET_MAX_PAYLOAD)
-			break;
 		hash = m0_pool_version2layout_id(pver, i);
 		l = m0_layout__list_lookup(dom, hash, true);
 		if (l != NULL) {
 			pa = &m0_layout_to_pdl(l)->pl_attr;
-			if (pa->pa_unit_size * pa->pa_N <= buffsize) {
-				lid = i;
-			} else {
-				m0_ref_put(&l->l_ref);
+			if (pa->pa_unit_size * pa->pa_N >= buffsize ||
+			    /*
+			     * Performance degrades for units bigger than
+			     * 4 * LNET_MAX_PAYLOAD (4MB atm), as tests show.
+			     */
+			    m0_lid_to_unit_map[i] > 4 * LNET_MAX_PAYLOAD)
 				break;
-			}
 			m0_ref_put(&l->l_ref);
 		}
 	}
 
-	hash = m0_pool_version2layout_id(pver, lid);
-	l = m0_layout__list_lookup(dom, hash, true);
 	if (l != NULL) {
-		pa = &m0_layout_to_pdl(l)->pl_attr;
 		M0_LOG(M0_INFO,
 			"Found lid=%d (pver+lid hash=%lx, unit_size=%d, N=%d) "
-			"by buffer size %d.", (int)lid, (unsigned long int)hash,
+			"by buffer size %d.", i, (unsigned long int)hash,
 			(int)pa->pa_unit_size, (int)pa->pa_N, (int)buffsize);
 		m0_ref_put(&l->l_ref);
 	}
 	m0_mutex_unlock(&dom->ld_lock);
 
-	return lid;
+	return (i == m0_lid_to_unit_map_nr) ? 0 : i;
```

Comment:
Please note, the previous version always return a valid lid (in the worst case, the M0_DEFAULT_LAYOUT_ID is returned). But now, it may return "0", which is not a valid lid.
Can we return M0_DEFAULT_LAYOUT_ID if (i == m0_lid_to_unit_map_nr) ?

Reply:
Yes, we can do that. But was thinking of returning an invalid lid (same as error) in this case. Default lid would hide the problem, I'm afraid. What do you think?

Reply:
I prefer to return the M0_DEFAULT_LAYOUT_ID in worst case.

Otherwise, we have to ensure that the caller checks the result and handle error properly in all places,
including to add comments about this.

Reply:
There are only 2 places where this function is called atm - I can update them too.

Reply:
Updated, please take a look.